---
title: "clustering_bartobar"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{clustering_SL_bartobar}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r libraries}

library('CAGEr')
library('ggplot2'); theme_set(theme_minimal())
library('BSgenome')
library('CAGErAid')

```

# Load in data from the [polishing](./vignettes/polishing.Rmd) file

```{r load in data}

ce <- readRDS('./data/BARtoBAR.rds')

```

```{r}

# load GFF files
gff <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2022-11-09_Barcelona_Oik/AlignWithRNAseqPipelinePE_BAR/Bar2_p4.gm.gtf')

```

# Clustering

Separate SL containing samples

```{r}

ce_sl <- ce[, ce$SLfound] |>
  normalizeTagCount(method = "simpleTpm") |>
  annotateCTSS(gff)

ce_no <- ce[, !ce$SLfound] |>
  normalizeTagCount(method = "simpleTpm") |>
  annotateCTSS(gff)

```

```{r plot reverse cumulatives - SL}

plotReverseCumulatives(ce_sl,
                       fitInRange = NULL,
                       values = "raw",
                       onePlot = TRUE)

plotReverseCumulatives(ce_sl,
                       fitInRange = NULL,
                       values = "normalized",
                       onePlot = TRUE)

```

```{r plot reverse cumulatives - no SL}

plotReverseCumulatives(ce_no,
                       fitInRange = NULL,
                       values = "raw",
                       onePlot = TRUE)

plotReverseCumulatives(ce_no,
                       fitInRange = NULL,
                       values = "normalized",
                       onePlot = TRUE)

```

```{r simple clustering and aggregating}

ce_sl <- quickClustersSL(ce_sl, aggregate = T, gff)
ce_no <- quickClustersSL(ce_no, aggregate = T, gff)

```

```{r export to bed - SL}

# # workaround with saving the scores beforehand
# scores <- score(consensusClustersGR(ce_sl))
# 
# # cctrack_sl <- exportToTrack(ce_sl, "consensusClusters")#, qLow = 0.1, qUp = 0.9)
# cctrack_sl <- consensusClustersGR(ce_sl, qL=.1, qU=.9, returnInterquantileWidth = TRUE) |>
#   exportToTrack(qL=.1, qU=.9)
# cctrack_sl$score <- scores
# 
# cctrack_sl@trackLine@description <- "CAGE Consensus Clusters for trans splicing sites"
# cctrack_sl@trackLine@name <- "sl"
# # Flat AG
# cctrack_sl$itemRgb <- ifelse(flagByUpstreamSequences(rowRanges(consensusClustersSE(ce_sl))$dominant_ctss, "AG"), "black", "grey")
# 
# cctrack_sl[cctrack_sl$itemRgb == "black"] |> score() |> decode() |> log10() |> hist(br=100)
# cctrack_sl[cctrack_sl$itemRgb == "grey"]  |> score() |> decode() |> log10() |> hist(br=100)
# 
# ####

cctrack_sl <- makeBed(ce_sl, sl_found = TRUE)

cctrack_sl[cctrack_sl$itemRgb == "black"] |> score() |> decode() |> log10() |> hist(br=100)
cctrack_sl[cctrack_sl$itemRgb == "grey"]  |> score() |> decode() |> log10() |> hist(br=100)

rtracklayer::export.bed(cctrack_sl, "./data/clusters_sl_bartobar.bed")

```

```{r export to bed - no SL}

cctrack_no <- makeBed(ce_no, sl_found = FALSE)

cctrack_no[cctrack_no$itemRgb == "black"] |> score() |> decode() |> log10() |> hist(br=100)
cctrack_no[cctrack_no$itemRgb == "grey"]  |> score() |> decode() |> log10() |> hist(br=100)

rtracklayer::export.bed(cctrack_no, "./data/clusters_no_bartobar.bed")

```
