---
---
title: "clustering_no"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{clustering_no}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r libraries}

library('CAGEr')
library('ggplot2'); theme_set(theme_bw())
library('CAGErAid')

```


# Load in data from the [polishing](./polishing.Rmd) file

```{r load in data}

ce <- SimpleList()

ce$BARtoBAR <- readRDS('./data/polished/BARtoBAR.rds')
ce$BARtoOKI <- readRDS('./data/polished/BARtoOKI.rds')
ce$BARtoOSA <- readRDS('./data/polished/BARtoOSA.rds')

ce$OKItoBAR <- readRDS('./data/polished/OKItoBAR.rds')
ce$OKItoOKI <- readRDS('./data/polished/OKItoOKI.rds')
ce$OKItoOSA <- readRDS('./data/polished/OKItoOSA.rds')

ce$OSAtoBAR <- readRDS('./data/polished/OSAtoBAR.rds')
ce$OSAtoOKI <- readRDS('./data/polished/OSAtoOKI.rds')
ce$OSAtoOSA <- readRDS('./data/polished/OSAtoOSA.rds')

```

```{r load gff files}

# load GFF files
gff_bar <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2022-11-09_Barcelona_Oik/AlignWithRNAseqPipelinePE_BAR/Bar2_p4.gm.gtf')
gff_oki <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2021-12-17_Okinawa_Oik/AlignWithRNAseqPipelinePE_OKI/Okinawa.genes.gtf')
gff_osa <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2022-02-09_Osaka_Oik/AlignWithRNAseqPipelinePE_OSA/OSKA2016v1.9.gm.gtf')

```

# Clustering

We separate samples without SL and check for power law distribution.

```{r plot reverse cumulatives}

for (i in names(ce)) {
  ce[[i]] <- ce[[i]][, !ce[[i]]$SLfound]
  plotReverseCumulatives(ce[[i]],
                       fitInRange = NULL,
                       values = "raw")
}

```

As seen in the reverse cumulative plot, samples without SL tend to follow the power law distribution, so we will use it for normalization. However, some 'toOSA' data does not have enough datapoints (?) and the algorithm crashes when using power law normalization. Therefore, in those cases, we will use simple TPM normalization.


```{r}

# for (i in names(ce)) {
#   ce[[i]] <- ce[[i]] |> normalizeTagCount(method = 'powerLaw') |> suppressWarnings()
  # plotReverseCumulatives(ce[[i]],
  #                        fitInRange = NULL,
  #                        values = "normalized",
  #                        onePlot = TRUE)
# 
  # }


ce$BARtoBAR <- normalizeTagCount(ce$BARtoBAR, 'powerLaw') |> suppressWarnings()
ce$BARtoOKI <- normalizeTagCount(ce$BARtoOKI, 'powerLaw') |> suppressWarnings()
ce$BARtoOSA <- normalizeTagCount(ce$BARtoOSA, 'simpleTpm') |> suppressWarnings()

ce$OKItoBAR <- normalizeTagCount(ce$OKItoBAR, 'powerLaw') |> suppressWarnings()
ce$OKItoOKI <- normalizeTagCount(ce$OKItoOKI, 'powerLaw') |> suppressWarnings()
ce$OKItoOSA <- normalizeTagCount(ce$OKItoOSA, 'simpleTpm') |> suppressWarnings()

ce$OSAtoBAR <- normalizeTagCount(ce$OSAtoBAR, 'powerLaw') |> suppressWarnings()
ce$OSAtoOKI <- normalizeTagCount(ce$OSAtoOKI, 'powerLaw') |> suppressWarnings()
ce$OSAtoOSA <- normalizeTagCount(ce$OSAtoOSA, 'powerLaw') |> suppressWarnings()


```

Now the samples are normalized and the reverse cumulative distribution roughly follows a straight line. Let's annotate the objects.

```{r annotate the CAGEexp object}

# annotate the CAGEexp objects
for (i in grep('toBAR', names(ce))) ce[[i]] <- annotateCTSS(ce[[i]], gff_bar) |> suppressWarnings()
for (i in grep('toOKI', names(ce))) ce[[i]] <- annotateCTSS(ce[[i]], gff_oki) |> suppressWarnings()
for (i in grep('toOSA', names(ce))) ce[[i]] <- annotateCTSS(ce[[i]], gff_osa) |> suppressWarnings()

```

Now we cluster and aggregate tags with the use of the distclu algorithm.

```{r simple clustering and aggregating}

for (i in grep('toBAR', names(ce))) ce[[i]] <- quickCC(ce[[i]], sl_found = FALSE, gff_bar)
for (i in grep('toOKI', names(ce))) ce[[i]] <- quickCC(ce[[i]], sl_found = FALSE, gff_oki)
for (i in grep('toOSA', names(ce))) ce[[i]] <- quickCC(ce[[i]], sl_found = FALSE, gff_osa)

```

Save the data: 

```{r saving clustered data}

saveRDS(ce$BARtoBAR, file = './data/clustered/BARtoBAR_no.rds')
saveRDS(ce$BARtoOKI, file = './data/clustered/BARtoOKI_no.rds')
saveRDS(ce$BARtoOSA, file = './data/clustered/BARtoOSA_no.rds')

saveRDS(ce$OKItoBAR, file = './data/clustered/OKItoBAR_no.rds')
saveRDS(ce$OKItoOKI, file = './data/clustered/OKItoOKI_no.rds')
saveRDS(ce$OKItoOSA, file = './data/clustered/OKItoOSA_no.rds')

saveRDS(ce$OSAtoBAR, file = './data/clustered/OSAtoBAR_no.rds')
saveRDS(ce$OSAtoOKI, file = './data/clustered/OSAtoOKI_no.rds')
saveRDS(ce$OSAtoOSA, file = './data/clustered/OSAtoOSA_no.rds')

```

And finally, export the data to bed format for track visualisation.

```{r export to bed}

for (i in names(ce)) {
  cctrack <- makeBed(ce[[i]], sl_found = FALSE)
  
  # cctrack[cctrack$itemRgb == "black"] |> score() |> decode() |> log10() |> hist(br =
  #                                                                                 100)
  # cctrack[cctrack$itemRgb == "grey"]  |> score() |> decode() |> log10() |> hist(br =
  #                                                                                 100)

  rtracklayer::export.bed(cctrack, paste0('./data/bed/clusters_no_', i, '.bed'))
  }
```
