---
title: "polishing"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{polishing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r libraries}
library('CAGEr')
library('ggplot2'); theme_set(theme_bw())
library('pheatmap')
library('devtools')
library('CAGErAid')
```

# Load in data from the [loading](./loading.Rmd) file

```{r load in data}

ce <- SimpleList()
ce$BARtoBAR <- readRDS('./data/BARtoBAR.rds')
ce$BARtoOKI <- readRDS('./data/BARtoOKI.rds')
ce$BARtoOSA <- readRDS('./data/BARtoOSA.rds')

ce$OKItoBAR <- readRDS('./data/OKItoBAR.rds')
ce$OKItoOKI <- readRDS('./data/OKItoOKI.rds')
ce$OKItoOSA <- readRDS('./data/OKItoOSA.rds')

ce$OSAtoBAR <- readRDS('./data/OSAtoBAR.rds')
ce$OSAtoOKI <- readRDS('./data/OSAtoOKI.rds')
ce$OSAtoOSA <- readRDS('./data/OSAtoOSA.rds')

```


# Adding Hisat2 and general QC stats to metadata

```{r hisat2 and general stats QC}

for (i in names(ce)) ce[[i]] <- quickMQC(ce[[i]])

```

After collecting statistics from the pipeline above, we can plot the data to see the levels of alignment.

```{r ploting mapping stats}

plotAnnot_ <- function(name)
  plotAnnot(ce[[name]], msScope_nfcore_rnaseq, group = 'RNA', facet = 'SLfactor', norm = FALSE) +
      ylab('Number of tags processed') +
      xlab('Sample name') +
      scale_y_continuous(guide = guide_axis(angle = 45)) +
      theme(plot.subtitle=element_text(size=8, face="italic")) +
      ggtitle(paste0(name, ': QC report of CAGE library alignment'),
              sub = 'The splice leader sequence was detected and removed before alignment')

for (name in names(ce)) plot(plotAnnot_(name))

```

Surprisingly, there is a large fraction of unmapped tags in the OKItoOKI and OSAtoOSA alignments, despite SL sequences being found. This could be due to the HMM algorithm not actually recognising the correct SL sequences, even a small amount of genetic distance causing tags not to be aligned, or artifacts in sequencing. Needs to be examined further.

Now we annotate the CTSSs and explore proportions of sequence features.

```{r annotating the CTSSs}

# load GFF files
gff_bar <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2022-11-09_Barcelona_Oik/AlignWithRNAseqPipelinePE_BAR/Bar2_p4.gm.gtf')
gff_oki <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2021-12-17_Okinawa_Oik/AlignWithRNAseqPipelinePE_OKI/Okinawa.genes.gtf')
gff_osa <- quickGFF('/bucket/LuscombeU/live/CharlesPlessy/CAGE/2022-02-09_Osaka_Oik/AlignWithRNAseqPipelinePE_OSA/OSKA2016v1.9.gm.gtf')


for (i in grep('toBAR', names(ce))) ce[[i]] <- annotateCTSS(ce[[i]], gff_bar) |> suppressWarnings()
for (i in grep('toOKI', names(ce))) ce[[i]] <- annotateCTSS(ce[[i]], gff_oki) |> suppressWarnings()
for (i in grep('toOSA', names(ce))) ce[[i]] <- annotateCTSS(ce[[i]], gff_osa) |> suppressWarnings()


# annotated results
for (i in names(ce)) {
  print(names(ce[i]))
  print(colData(ce[[i]])[, c('librarySizes', 'promoter', 'exon', 'intron', 'unknown')])}

# proportions
for (i in names(ce)) {
  print(names(ce[i]))
  print(colData(ce[[i]])[,c('promoter', 'exon', 'intron', 'unknown')] |> as.matrix() |> prop.table(1)  |> round(2))
        }

```
Now we can plot the annotated results to see the proportions of sequence features present in the data. 

```{r plot annotated results - normalized counts}

plotAnnot_ <- function(name) {
  plotAnnot(ce[[name]], 'counts', group = 'RNA', facet = 'SLfactor', normalise = TRUE) +
      ylab('Fraction of tags aligned') +
      xlab('Sample name') +
      scale_y_continuous(guide = guide_axis(angle = 45)) +
      theme(plot.subtitle=element_text(size=8, face="italic")) +
      ggtitle(paste0(name, ': Annotation of the aligned tags'),
              sub = 'The promoters are defined as the 1-kb window centered on the transcript start site.')}

for (name in names(ce)) plot(plotAnnot_(name))

```

The high proportions of promoters in SL data is due to SL signals being positioned close to ATG. The regions where SL is found can be considered to be more highly conserved. We can also explore the total tag count on the OSAtoOSA alignment.

To check for mislabeling, we can roughly cluster the number of aligned tags per chromosome:

```{r correlate expression per chromosome}

roughClusters <- function(name) {
  seqNameTotalsSE(ce[[name]]) |> assay() |> cor(meth='spe') |> pheatmap::pheatmap(main = name)
}

# sanity check
roughClusters('BARtoBAR')
roughClusters('OKItoOKI')
roughClusters('OSAtoOSA')

```

There are no suspicious outliers in the plots and the expression patterns are clustered together in SL samples and noSL samples correctly. 

Plotting the reverse cumulative distribution reveals differences between the samples with and without SL found:

```{r reverse cumulative distribution of the CTSSes}

colors_bar <- c("red", "red", "red",  "red", "blue",  "blue" ,"pink", "pink"  ,"pink","pink",  "green", "green",   "green", "green",  "green",  "green",  "green" ,  "green" )
colors_oki <- c("red", "red", "pink", "pink", "blue", "blue", "lightblue", "lightblue","green", "green")
colors_osa <- c("purple", "purple", "red", "red", "pink", "pink", "blue", "blue", "lightblue", "lightblue", "green", "green")


for (i in grep('BARto', names(ce))) ce[[i]] <- setColors(ce[[i]], colors_bar)
for (i in grep('OKIto', names(ce))) ce[[i]] <- setColors(ce[[i]], colors_oki)
for (i in grep('OSAto', names(ce))) ce[[i]] <- setColors(ce[[i]], colors_osa)



plotReverseCumulatives(ce$BARtoBAR, fitInRange = c(1e3, 1e5), values = "raw")
plotReverseCumulatives(ce$OKItoBAR, fitInRange = c(1e3, 1e5), values = "raw")
plotReverseCumulatives(ce$OKItoOKI, fitInRange = c(1e3, 1e5), values = "raw")
plotReverseCumulatives(ce$OSAtoBAR, fitInRange = c(1e3, 1e5), values = "raw")
plotReverseCumulatives(ce$OSAtoOSA, fitInRange = c(1e3, 1e5), values = "raw")

```

Some of the above cross-alignments couldn't be plotted due to an insufficient data. For the remaining cross-alignments, we see that the distribution in some samples (without SL) follows a more diagonal, straight line, and in some (SL found) the line of the distribution is more curved, not following a power law distribution. We can see that the two parts of the data differ, and should not be analysed together. Therefore, in the further steps, we will analyse SL and no-SL data separately.


```{r exporting unclustered data to bed or bw - maybe, eval=FALSE, include=FALSE}

# bed
for (i in names(ce)) {
  ce[[i]] <- normalizeTagCount(ce[[i]])
  trks <- exportToTrack(ce[[i]], oneTrack = FALSE)
  for (n in seq_along(trks)) {
    name <- sampleLabels(ce[[i]])[n]
    rtracklayer::export.bed(trks[n], paste0("./data/bed/", name, ".ctss.bed.gz"))
  }

# bigWig
  for (n in seq_along(trks)) {
    name <- sampleLabels(ce[[i]])[n]
    trkL <- split(trks[[n]], strand(trks[[n]]), drop = TRUE)
    trkL[['+']]@trackLine@description <- paste(name, " plus")
    trkL[['-']]@trackLine@description <- paste(name, " plus")
    rtracklayer::export.bw(trkL[['+']], paste0("./data/bw/", name, ".plus.bw"))
    rtracklayer::export.bw(trkL[['-']], paste0("./data/bw/", name, ".minus.bw"))
  }
}

```

Finally, we save the CAGEexp objects for the next steps of analysis.

```{r saving modified CAGEexp objects}

saveRDS(ce$BARtoBAR, file = './data/polished/BARtoBAR.rds')
saveRDS(ce$BARtoOKI, file = './data/polished/BARtoOKI.rds')
saveRDS(ce$BARtoOSA, file = './data/polished/BARtoOSA.rds')

saveRDS(ce$OKItoBAR, file = './data/polished/OKItoBAR.rds')
saveRDS(ce$OKItoOKI, file = './data/polished/OKItoOKI.rds')
saveRDS(ce$OKItoOSA, file = './data/polished/OKItoOSA.rds')

saveRDS(ce$OSAtoBAR, file = './data/polished/OSAtoBAR.rds')
saveRDS(ce$OSAtoOKI, file = './data/polished/OSAtoOKI.rds')
saveRDS(ce$OSAtoOSA, file = './data/polished/OSAtoOSA.rds')

```
